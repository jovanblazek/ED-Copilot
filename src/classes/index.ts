export * from './Error'
export * from './Tick'
