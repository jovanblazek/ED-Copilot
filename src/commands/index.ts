import { map, reduce } from 'lodash'
import commanderProfile from './commanderProfile'
import faction from './faction'
import interstellarFactors from './interstellarFactors'
import materialTrader from './materialTrader'
import ping from './ping'
import setup from './setup'
import systemDistance from './systemDistance'
import systemInfo from './systemInfo'
import techBroker from './techBroker'
import tick from './tick'
import { CommandHandlerArgs } from './types'

type CommandHandlersType = {
  [key: string]: (args: CommandHandlerArgs) => Promise<void>
}

export const Commands = {
  ping,
  techBroker,
  materialTrader,
  interstellarFactors,
  systemDistance,
  tick,
  systemInfo,
  faction,
  commanderProfile,
  setup,
}

export const CommandBuilders = map(Commands, (command) => command.builder.toJSON())

export const CommandHandlers: CommandHandlersType = reduce(
  Commands,
  (commandHandlers, command) => ({
    ...commandHandlers,
    [command.builder.name]: command.handler,
  }),
  {}
)
