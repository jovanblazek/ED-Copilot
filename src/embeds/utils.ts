const MAX_EMBED_DESCRIPTION_LENGTH = 3800 // Added some buffer to be safe

/**
 * Splits text into chunks that fit within Discord embed description limits
 * Useful when you need to use markdown, but embed field titles do not support it
 */
export const chunkDescription = (
  lines: string[],
  maxLength: number = MAX_EMBED_DESCRIPTION_LENGTH
): string[] => {
  const chunks: string[] = []
  let currentChunk = ''

  for (const line of lines) {
    const lineWithNewlines = `${line}\n\n`

    if (currentChunk.length + lineWithNewlines.length > maxLength) {
      if (currentChunk) {
        chunks.push(currentChunk.trim())
        currentChunk = ''
      }
    }

    currentChunk += lineWithNewlines
  }

  if (currentChunk) {
    chunks.push(currentChunk.trim())
  }

  return chunks
}
