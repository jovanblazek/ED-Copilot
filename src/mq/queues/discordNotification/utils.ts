import type { GuildFaction } from '@prisma/client'
import type { Client } from 'discord.js'
import { ChannelType } from 'discord.js'

export const getNotificationChannelFromGuildFactionOrThrow = ({
  client,
  guildFaction,
}: {
  client: Client
  guildFaction: GuildFaction
}) => {
  const { notificationChannelId } = guildFaction
  if (!notificationChannelId) {
    throw new Error('Guild faction has no notification channel set')
  }

  const channel = client.channels.cache.get(notificationChannelId)
  if (
    !channel ||
    !(channel.type === ChannelType.GuildAnnouncement || channel.type === ChannelType.GuildText)
  ) {
    throw new Error(`Invalid notification channel, or type: ${channel?.type}`)
  }

  return channel
}
