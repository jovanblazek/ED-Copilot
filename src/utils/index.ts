export * from './embed'
export * from './pagination'
export * from './registerCommands'
export * from './translations'
