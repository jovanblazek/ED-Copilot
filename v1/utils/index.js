const { init } = require('./init')
const { getRandomActivity } = require('./activityChanger')
const { validateArgs } = require('./arguments')
const { parseSystemName } = require('./systemName')
const { fetchTickTime, wasAfterTick } = require('./tick')
const { createEmbed } = require('./embed')
const { tickDetector } = require('./tickDetector')
const { getCommandSyntax, getArgumentInfo, getArgumentOptions } = require('./commadSyntax')
const { getStationIcon, addSuffixToInt, printTrend } = require('./helpers')

module.exports = {
	createEmbed,
	init,
	getArgumentInfo,
	getArgumentOptions,
	getCommandSyntax,
	getRandomActivity,
	fetchTickTime,
	parseSystemName,
	tickDetector,
	validateArgs,
	wasAfterTick,
	getStationIcon,
	addSuffixToInt,
	printTrend,
}
