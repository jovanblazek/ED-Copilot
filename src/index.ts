import { Client, Intents } from 'discord.js'
import { timezone } from '../config.json'
import { Tick } from './classes'
import { CommandHandlers } from './commands'
import { errorHandler, fetchTickTime, initTranslations } from './utils'
import logger from './utils/logger'
import initTickDetector from './utils/tickDetector'
import './utils/environment'

const { BOT_TOKEN } = process.env

const BotClient = new Client({ intents: [Intents.FLAGS.GUILDS, Intents.FLAGS.GUILD_MESSAGES] })
const SavedTick = new Tick(null, timezone)

BotClient.once('ready', async () => {
  await fetchTickTime().then((tickTime) => {
    if (tickTime) {
      SavedTick.setTicktime(tickTime)
    }
  })
  await initTranslations()
  initTickDetector(BotClient, SavedTick)

  logger.info('Bot is ready!')
})

BotClient.on('interactionCreate', async (interaction) => {
  if (!interaction.isCommand()) {
    return
  }

  const { commandName } = interaction
  const handler = CommandHandlers[commandName]

  try {
    if (handler) {
      await CommandHandlers[commandName](interaction, SavedTick)
    }
  } catch (error) {
    await errorHandler(error, interaction)
  }
})

void BotClient.login(BOT_TOKEN)
