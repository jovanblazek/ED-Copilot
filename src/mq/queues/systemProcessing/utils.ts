import type { FactionState, Prisma as PrismaClientType } from '@prisma/client'
import { StateType } from '@prisma/client'
import { RedisKeys } from '../../../constants'
import type { EDDNConflict, EDDNFaction, EDDNFactionState } from '../../../types/eddn'
import type { TrackedFaction } from '../../../types/redis'
import logger from '../../../utils/logger'
import { getTrackedFactions, Redis } from '../../../utils/redis'
import type { Conflict } from '../discordNotification/types'
import { CONFLICT_STATES } from './constants'
import type { StateChanges } from './types'

export const getTrackedFactionsInSystem = async (eventFactions: EDDNFaction[]) => {
  const trackedFactions = await getTrackedFactions()
  return trackedFactions.filter((factionFromRedis) =>
    eventFactions.find(({ Name }) => Name === factionFromRedis.name)
  )
}

const getStatesToEnd = (currentDbStates: FactionState[], statesFromEvent: EDDNFactionState[]) =>
  // eslint-disable-next-line @typescript-eslint/no-unsafe-enum-comparison
  currentDbStates.filter((state) => !statesFromEvent.some((s) => s.State === state.stateName))

const getStatesToStart = (currentDbStates: FactionState[], statesFromEvent: EDDNFactionState[]) =>
  // eslint-disable-next-line @typescript-eslint/no-unsafe-enum-comparison
  statesFromEvent.filter((state) => !currentDbStates.some((s) => s.stateName === state.State))

export const getAllStatesToEnd = ({
  currentDbStatesByType,
  factionFromEvent,
}: {
  currentDbStatesByType: Record<StateType, FactionState[]>
  factionFromEvent: EDDNFaction
}): FactionState[] => {
  const activeStatesFromEvent = factionFromEvent?.ActiveStates ?? []
  const pendingStatesFromEvent = factionFromEvent?.PendingStates ?? []
  const recoveringStatesFromEvent = factionFromEvent?.RecoveringStates ?? []

  // End states that are no longer active
  const activeStatesToEnd = getStatesToEnd(
    currentDbStatesByType[StateType.Active],
    activeStatesFromEvent
  )
  const pendingStatesToEnd = getStatesToEnd(
    currentDbStatesByType[StateType.Pending],
    pendingStatesFromEvent
  )
  const recoveringStatesToEnd = getStatesToEnd(
    currentDbStatesByType[StateType.Recovering],
    recoveringStatesFromEvent
  )

  return [...activeStatesToEnd, ...pendingStatesToEnd, ...recoveringStatesToEnd]
}

export const getAllStatesToStart = ({
  currentDbStatesByType,
  factionFromEvent,
}: {
  currentDbStatesByType: Record<StateType, FactionState[]>
  factionFromEvent: EDDNFaction
}) => {
  const activeStatesFromEvent = factionFromEvent?.ActiveStates ?? []
  const pendingStatesFromEvent = factionFromEvent?.PendingStates ?? []
  const recoveringStatesFromEvent = factionFromEvent?.RecoveringStates ?? []

  const activeStatesToStart = getStatesToStart(
    currentDbStatesByType[StateType.Active],
    activeStatesFromEvent
  )
  const pendingStatesToStart = getStatesToStart(
    currentDbStatesByType[StateType.Pending],
    pendingStatesFromEvent
  )
  const recoveringStatesToStart = getStatesToStart(
    currentDbStatesByType[StateType.Recovering],
    recoveringStatesFromEvent
  )

  return {
    activeStatesToStart,
    pendingStatesToStart,
    recoveringStatesToStart,
  }
}

export const handleStateChanges = async (
  trx: PrismaClientType.TransactionClient,
  {
    trackedFaction,
    factionFromEvent,
    systemName,
    timestamp,
    currentDbStatesByType,
  }: {
    trackedFaction: TrackedFaction
    factionFromEvent: EDDNFaction
    systemName: string
    timestamp: string
    currentDbStatesByType: Record<StateType, FactionState[]>
  }
): Promise<StateChanges> => {
  const statesToEnd = getAllStatesToEnd({
    currentDbStatesByType,
    factionFromEvent,
  })

  if (statesToEnd.length > 0) {
    logger.info(
      statesToEnd,
      `systemProcessingWorker: ${systemName} - ${trackedFaction.name} - ENDING STATES`
    )

    await trx.factionState.updateMany({
      where: { id: { in: statesToEnd.map((state) => state.id) } },
      data: { endedAt: new Date(timestamp) },
    })
  }

  const { activeStatesToStart, pendingStatesToStart, recoveringStatesToStart } =
    getAllStatesToStart({
      factionFromEvent,
      currentDbStatesByType,
    })

  logger.info(
    {
      activeStatesToStart,
      pendingStatesToStart,
      recoveringStatesToStart,
    },
    `systemProcessingWorker: ${systemName} - ${trackedFaction.name} - STARTING STATES`
  )

  const createStateData = (
    states: EDDNFactionState[],
    stateType: StateType
  ): PrismaClientType.FactionStateCreateManyInput[] =>
    states.map((state) => ({
      factionId: trackedFaction.id,
      systemName,
      stateName: state.State,
      stateType,
      startedAt: new Date(timestamp),
    }))

  await trx.factionState.createMany({
    data: [
      ...createStateData(activeStatesToStart, StateType.Active),
      ...createStateData(pendingStatesToStart, StateType.Pending),
      ...createStateData(recoveringStatesToStart, StateType.Recovering),
    ],
  })

  return {
    statesToEnd,
    activeStatesToStart,
    pendingStatesToStart,
    recoveringStatesToStart,
  }
}

export const isConflictInEDDNStateArray = (stateArray: EDDNFactionState[]) =>
  stateArray.some((state) => CONFLICT_STATES.includes(state.State))

export const getConflictByFactionName = (conflicts: EDDNConflict[], factionName: string) =>
  conflicts.find(
    (conflict) => conflict.Faction1.Name === factionName || conflict.Faction2.Name === factionName
  )

export const transformConflictToDiscordNotificationData = (conflict: EDDNConflict): Conflict => ({
  faction1: {
    name: conflict.Faction1.Name,
    stake: conflict.Faction1.Stake,
    wonDays: conflict.Faction1.WonDays,
  },
  faction2: {
    name: conflict.Faction2.Name,
    stake: conflict.Faction2.Stake,
    wonDays: conflict.Faction2.WonDays,
  },
  status: conflict.Status,
  conflictType: conflict.WarType,
})

export const isSystemAlreadyProcessed = ({
  systemName,
  tickTimeISO,
}: {
  systemName: string
  tickTimeISO: string
}) => {
  const processedSystemRedisKey = RedisKeys.processedSystem({
    tickTimestamp: tickTimeISO,
    systemName,
  })
  return Redis.exists(processedSystemRedisKey)
}

export const groupFactionStatesByType = (factionStates: FactionState[]) =>
  factionStates.reduce(
    (acc, state) => {
      acc[state.stateType] = acc[state.stateType] || []
      acc[state.stateType].push(state)
      return acc
    },
    {
      [StateType.Active]: [] as FactionState[],
      [StateType.Pending]: [] as FactionState[],
      [StateType.Recovering]: [] as FactionState[],
    }
  )
