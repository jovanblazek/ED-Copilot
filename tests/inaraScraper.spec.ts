import { getInaraData } from '../src/utils'

const NO_RECORDS_URL =
  'https://inara.cz/nearest-stations/?ps1=sol&pi13=&pi14=0&pi15=0&pi16=&pi1=0&pi18=0&pi19=2000&pi17=0&ps2=&pi8=&pi9=0&pa2%5B20%5D=1&pi3=&pi4=0&pi5=0&pi6=0&pi7=0&pi23=0'

const URLS = {
  factors: {
    url: 'https://inara.cz/nearest-stations/?ps1=sol&pi13=&pi14=0&pi15=0&pi16=&pi1=0&pi18=3&pi19=2000&pi17=1&pi2=1&pa1%5B18%5D=1&pi8=&pi9=0&pi3=&pi4=0&pi5=0&pi6=0&pi7=0&pi23=0',
    cellsPerRow: 7,
  },
  broker: {
    url: 'https://inara.cz/nearest-stations/?ps1=sol&pi13=&pi14=0&pi15=0&pi16=&pi1=0&pi18=0&pi19=2000&pa1%5B26%5D=1&pi8=&pi9=0&pi3=&pi4=0&pi5=0&pi6=0&pi7=0&pi23=0',
    cellsPerRow: 8,
  },
  trader: {
    url: 'https://inara.cz/nearest-stations/?ps1=sol&pi13=&pi14=0&pi15=0&pi16=&pi1=0&pi18=0&pi19=2000&pa1%5B25%5D=1&pi8=&pi9=0&pi3=&pi4=0&pi5=0&pi6=0&pi7=0&pi23=0',
    cellsPerRow: 8,
  },
}

describe('Inara scraper', () => {
  it('should return empty array if no data is found', async () => {
    const data = await getInaraData(NO_RECORDS_URL, 8)
    expect(data.length).toBe(0)
    expect(data).toEqual([])
  })

  it('should return broker data', async () => {
    const data = await getInaraData(URLS.broker.url, URLS.broker.cellsPerRow)
    expect(data.length).toBeGreaterThan(0)
    expect(data[0]).toMatchSnapshot(
      {
        distanceLs: expect.stringContaining('Ls'),
        distanceLy: expect.stringContaining('Ly'),
      },
      'Tech broker data snapshot'
    )
  })

  it('should return trader data', async () => {
    const data = await getInaraData(URLS.trader.url, URLS.trader.cellsPerRow)
    expect(data.length).toBeGreaterThan(0)
    expect(data[0]).toMatchSnapshot(
      {
        distanceLs: expect.stringContaining('Ls'),
        distanceLy: expect.stringContaining('Ly'),
      },
      'Material trader data snapshot'
    )
  })

  // Interstellar factors avaliablity changes over time due to BGS
  it('should return factors data', async () => {
    const data = await getInaraData(URLS.factors.url, URLS.factors.cellsPerRow)
    expect(data.length).toBeGreaterThan(0)
    expect(data[0]).toMatchSnapshot(
      {
        station: expect.any(String),
        system: expect.any(String),
        distanceLs: expect.stringContaining('Ls'),
        distanceLy: expect.stringContaining('Ly'),
      },
      'Interstellar factors data snapshot'
    )
  })
})
