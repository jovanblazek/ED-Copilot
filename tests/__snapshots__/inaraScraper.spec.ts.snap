// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Inara scraper should return broker data: Tech broker data snapshot 1`] = `
Object {
  "distanceLs": StringContaining "Ls",
  "distanceLy": StringContaining "Ly",
  "station": "Dobrovolskiy Enterprise",
  "system": "WISE 1506+7027",
  "type": "Human",
}
`;

exports[`Inara scraper should return factors data: Interstellar factors data snapshot 1`] = `
Object {
  "distanceLs": StringContaining "Ls",
  "distanceLy": StringContaining "Ly",
  "station": Any<String>,
  "system": Any<String>,
  "type": null,
}
`;

exports[`Inara scraper should return trader data: Material trader data snapshot 1`] = `
Object {
  "distanceLs": StringContaining "Ls",
  "distanceLy": StringContaining "Ly",
  "station": "Patterson Enterprise",
  "system": "Sirius",
  "type": "Manufactured",
}
`;
